File: src/shared/services/fs-utils.ts
/**
 * TypeScript client for fs_utils Tauri commands
 *
 * This module provides type-safe access to all filesystem operations
 * exposed by the Rust backend via Tauri commands.
 */

import { invoke } from '@tauri-apps/api/core'

// ============================================================================
// Type Definitions
// ============================================================================

export interface FileMetadata {
  path: string
  is_dir: boolean
  is_file: boolean
  size: number // bytes
  modified: number // Unix timestamp in seconds
}

export interface DirEntry {
  name: string
  path: string
  is_dir: boolean
}

// ============================================================================
// Core File Operations
// ============================================================================

/**
 * Read entire file as UTF-8 string
 */
export async function readFile(path: string): Promise<string> {
  return invoke<string>('read_file', { path })
}

/**
 * Write string to file (creates file if it doesn't exist, overwrites if it does)
 */
export async function writeFile(path: string, content: string): Promise<void> {
  return invoke<void>('write_file', { path, content })
}

/**
 * Read file as bytes
 */
export async function readFileBytes(path: string): Promise<number[]> {
  return invoke<number[]>('read_file_bytes', { path })
}

/**
 * Get file metadata (size, modified time, etc.)
 */
export async function getFileStat(path: string): Promise<FileMetadata> {
  return invoke<FileMetadata>('get_file_stat', { path })
}

// ============================================================================
// Directory Operations
// ============================================================================

/**
 * List directory contents
 * Returns sorted array of directory entries
 */
export async function readDir(path: string): Promise<DirEntry[]> {
  return invoke<DirEntry[]>('read_dir', { path })
}

/**
 * Check if file or directory exists
 */
export async function fileExists(path: string): Promise<boolean> {
  return invoke<boolean>('file_exists', { path })
}

/**
 * Create directory (creates parent directories as needed)
 */
export async function createDir(path: string): Promise<void> {
  return invoke<void>('create_dir', { path })
}

/**
 * Delete file
 */
export async function deleteFile(path: string): Promise<void> {
  return invoke<void>('delete_file', { path })
}

/**
 * Delete directory (recursively removes all contents)
 */
export async function deleteDir(path: string): Promise<void> {
  return invoke<void>('delete_dir', { path })
}

/**
 * Copy file from src to dest
 */
export async function copyFile(src: string, dest: string): Promise<void> {
  return invoke<void>('copy_file', { src, dest })
}

/**
 * Rename/move file
 */
export async function renameFile(from: string, to: string): Promise<void> {
  return invoke<void>('rename_file', { from, to })
}

// ============================================================================
// JSONL File Operations (Optimized for Large Transcripts)
// ============================================================================

/**
 * Read JSONL file (newline-delimited JSON) as array of JSON objects
 * Optimized for large files - reads line by line
 *
 * Example file:
 * {"id": 1, "name": "foo"}
 * {"id": 2, "name": "bar"}
 *
 * Returns: [{"id": 1, "name": "foo"}, {"id": 2, "name": "bar"}]
 */
export async function readJsonl(path: string): Promise<Record<string, unknown>[]> {
  const lines = await invoke<Record<string, unknown>[]>('read_jsonl', { path })
  return lines
}

/**
 * Read JSONL file with pagination
 * Useful for large transcripts - returns only a specific range of lines
 *
 * @param path - Path to JSONL file
 * @param offset - Starting line number
 * @param limit - Maximum number of lines to return
 */
export async function readJsonlPaginated(
  path: string,
  offset: number,
  limit: number,
): Promise<Record<string, unknown>[]> {
  const lines = await invoke<Record<string, unknown>[]>('read_jsonl_paginated', {
    path,
    offset,
    limit,
  })
  return lines
}

// ============================================================================
// Path Operations
// ============================================================================

/**
 * Get the user's home directory
 * E.g., /Users/dinokljuco on macOS, C:\Users\dinokljuco on Windows
 */
export async function getHomeDir(): Promise<string> {
  return invoke<string>('get_home_dir', {})
}

/**
 * Join path segments
 * Platform-aware (uses / on Unix, \ on Windows)
 */
export async function pathJoin(parts: string[]): Promise<string> {
  return invoke<string>('path_join', { parts })
}

/**
 * Get directory name from a path
 * E.g., /home/user/file.txt -> /home/user
 */
export async function pathDirname(path: string): Promise<string> {
  return invoke<string>('path_dirname', { path })
}

/**
 * Get file name from a path
 * E.g., /home/user/file.txt -> file.txt
 */
export async function pathBasename(path: string): Promise<string> {
  return invoke<string>('path_basename', { path })
}

/**
 * Normalize a path (resolve . and .., etc.)
 */
export async function pathNormalize(path: string): Promise<string> {
  return invoke<string>('path_normalize', { path })
}

// ============================================================================
// Convenience Functions
// ============================================================================

/**
 * Read JSON file and parse it
 */
export async function readJson<T = unknown>(path: string): Promise<T> {
  const content = await readFile(path)
  return JSON.parse(content) as T
}

/**
 * Write object as JSON file
 */
export async function writeJson<T = unknown>(path: string, data: T): Promise<void> {
  const content = JSON.stringify(data, null, 2)
  return writeFile(path, content)
}

/**
 * Build a path and ensure it exists
 * E.g., buildPath(['~', '.claude', 'projects'])
 */
export async function buildPath(parts: string[]): Promise<string> {
  // Handle ~ expansion
  let expandedParts = parts
  if (parts[0] === '~') {
    const home = await getHomeDir()
    expandedParts = [home, ...parts.slice(1)]
  }

  return pathJoin(expandedParts)
}

File: src-tauri/src/commands/fs_utils.rs
/**
 * Core file system utilities for all Tauri commands
 *
 * This module provides the foundation for all file operations:
 * - Reading/writing files
 * - Directory operations
 * - JSONL file handling (optimized for large transcript files)
 * - Path resolution
 */

use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use tokio::fs;
use tokio::io::{AsyncBufReadExt, BufReader};

// ============================================================================
// Type Definitions
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileMetadata {
    pub path: String,
    pub is_dir: bool,
    pub is_file: bool,
    pub size: u64,
    pub modified: u64, // Unix timestamp
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DirEntry {
    pub name: String,
    pub path: String,
    pub is_dir: bool,
}

// ============================================================================
// Core File Operations
// ============================================================================

/// Read entire file as string
#[tauri::command]
pub async fn read_file(path: String) -> Result<String, String> {
    fs::read_to_string(&path)
        .await
        .map_err(|e| format!("Failed to read file '{}': {}", path, e))
}

/// Write string to file
#[tauri::command]
pub async fn write_file(path: String, content: String) -> Result<(), String> {
    // Ensure parent directory exists
    if let Some(parent) = PathBuf::from(&path).parent() {
        if !parent.as_os_str().is_empty() {
            fs::create_dir_all(parent)
                .await
                .map_err(|e| format!("Failed to create parent directories: {}", e))?;
        }
    }

    fs::write(&path, &content)
        .await
        .map_err(|e| format!("Failed to write file '{}': {}", path, e))
}

/// Read file as bytes
#[tauri::command]
pub async fn read_file_bytes(path: String) -> Result<Vec<u8>, String> {
    fs::read(&path)
        .await
        .map_err(|e| format!("Failed to read file '{}': {}", path, e))
}

/// Get file metadata
#[tauri::command]
pub async fn get_file_stat(path: String) -> Result<FileMetadata, String> {
    let metadata = fs::metadata(&path)
        .await
        .map_err(|e| format!("Failed to stat file '{}': {}", path, e))?;

    Ok(FileMetadata {
        path: path.clone(),
        is_dir: metadata.is_dir(),
        is_file: metadata.is_file(),
        size: metadata.len(),
        modified: metadata
            .modified()
            .ok()
            .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
            .map(|d| d.as_secs())
            .unwrap_or(0),
    })
}

// ============================================================================
// Directory Operations
// ============================================================================

/// List directory contents
#[tauri::command]
pub async fn read_dir(path: String) -> Result<Vec<DirEntry>, String> {
    let mut entries = Vec::new();
    let mut dir = fs::read_dir(&path)
        .await
        .map_err(|e| format!("Failed to read directory '{}': {}", path, e))?;

    while let Some(entry) = dir
        .next_entry()
        .await
        .map_err(|e| format!("Error reading directory entries: {}", e))?
    {
        let metadata = entry
            .metadata()
            .await
            .map_err(|e| format!("Failed to get metadata: {}", e))?;

        let file_name = entry.file_name();
        let name = file_name.to_string_lossy().to_string();
        let entry_path = entry.path();
        let entry_path_str = entry_path.to_string_lossy().to_string();

        entries.push(DirEntry {
            name,
            path: entry_path_str,
            is_dir: metadata.is_dir(),
        });
    }

    // Sort by name for consistent results
    entries.sort_by(|a, b| a.name.cmp(&b.name));

    Ok(entries)
}

/// Check if file/directory exists
#[tauri::command]
pub async fn file_exists(path: String) -> Result<bool, String> {
    match fs::try_exists(&path).await {
        Ok(exists) => Ok(exists),
        Err(e) => Err(format!("Failed to check if file exists: {}", e)),
    }
}

/// Create directory
#[tauri::command]
pub async fn create_dir(path: String) -> Result<(), String> {
    fs::create_dir_all(&path)
        .await
        .map_err(|e| format!("Failed to create directory '{}': {}", path, e))
}

/// Delete file
#[tauri::command]
pub async fn delete_file(path: String) -> Result<(), String> {
    fs::remove_file(&path)
        .await
        .map_err(|e| format!("Failed to delete file '{}': {}", path, e))
}

/// Delete directory (recursive)
#[tauri::command]
pub async fn delete_dir(path: String) -> Result<(), String> {
    fs::remove_dir_all(&path)
        .await
        .map_err(|e| format!("Failed to delete directory '{}': {}", path, e))
}

/// Copy file
#[tauri::command]
pub async fn copy_file(src: String, dest: String) -> Result<(), String> {
    // Ensure parent directory exists
    if let Some(parent) = PathBuf::from(&dest).parent() {
        if !parent.as_os_str().is_empty() {
            fs::create_dir_all(parent)
                .await
                .map_err(|e| format!("Failed to create parent directories: {}", e))?;
        }
    }

    fs::copy(&src, &dest)
        .await
        .map_err(|e| format!("Failed to copy file from '{}' to '{}': {}", src, dest, e))?;

    Ok(())
}

/// Rename/move file
#[tauri::command]
pub async fn rename_file(from: String, to: String) -> Result<(), String> {
    fs::rename(&from, &to)
        .await
        .map_err(|e| format!("Failed to rename file from '{}' to '{}': {}", from, to, e))
}

// ============================================================================
// JSONL File Operations (Optimized for Large Transcripts)
// ============================================================================

/// Read JSONL file (newline-delimited JSON) as array of JSON objects
/// This is optimized for large files by reading line-by-line
#[tauri::command]
pub async fn read_jsonl(path: String) -> Result<Vec<serde_json::Value>, String> {
    let file = fs::File::open(&path)
        .await
        .map_err(|e| format!("Failed to open file '{}': {}", path, e))?;

    let reader = BufReader::new(file);
    let mut lines = reader.lines();
    let mut results = Vec::new();

    while let Some(line) = lines
        .next_line()
        .await
        .map_err(|e| format!("Error reading line: {}", e))?
    {
        if line.trim().is_empty() {
            continue;
        }

        match serde_json::from_str::<serde_json::Value>(&line) {
            Ok(json) => results.push(json),
            Err(e) => {
                // Log parsing error but continue processing
                log::warn!("Failed to parse JSONL line: {}", e);
            }
        }
    }

    Ok(results)
}

/// Read JSONL file with pagination (for large files)
/// Returns a specific range of lines
#[tauri::command]
pub async fn read_jsonl_paginated(
    path: String,
    offset: usize,
    limit: usize,
) -> Result<Vec<serde_json::Value>, String> {
    let file = fs::File::open(&path)
        .await
        .map_err(|e| format!("Failed to open file '{}': {}", path, e))?;

    let reader = BufReader::new(file);
    let mut lines = reader.lines();
    let mut results = Vec::new();
    let mut current_line = 0;

    while let Some(line) = lines
        .next_line()
        .await
        .map_err(|e| format!("Error reading line: {}", e))?
    {
        if line.trim().is_empty() {
            continue;
        }

        if current_line >= offset && current_line < offset + limit {
            match serde_json::from_str::<serde_json::Value>(&line) {
                Ok(json) => results.push(json),
                Err(e) => {
                    log::warn!("Failed to parse JSONL line: {}", e);
                }
            }
        }

        if current_line >= offset + limit {
            break;
        }

        current_line += 1;
    }

    Ok(results)
}

// ============================================================================
// Path Operations
// ============================================================================

/// Get the user's home directory
#[tauri::command]
pub fn get_home_dir() -> Result<String, String> {
    dirs::home_dir()
        .ok_or_else(|| "Could not determine home directory".to_string())
        .map(|p| p.to_string_lossy().to_string())
}

/// Join path segments
#[tauri::command]
pub fn path_join(parts: Vec<String>) -> Result<String, String> {
    if parts.is_empty() {
        return Err("No path parts provided".to_string());
    }

    let mut path = PathBuf::from(&parts[0]);
    for part in &parts[1..] {
        path.push(part);
    }

    Ok(path.to_string_lossy().to_string())
}

/// Get the directory name from a path
#[tauri::command]
pub fn path_dirname(path: String) -> Result<String, String> {
    PathBuf::from(&path)
        .parent()
        .ok_or_else(|| "No parent directory".to_string())
        .map(|p| p.to_string_lossy().to_string())
}

/// Get the file name from a path
#[tauri::command]
pub fn path_basename(path: String) -> Result<String, String> {
    PathBuf::from(&path)
        .file_name()
        .ok_or_else(|| "No file name in path".to_string())
        .map(|n| n.to_string_lossy().to_string())
}

/// Normalize a path
#[tauri::command]
pub fn path_normalize(path: String) -> Result<String, String> {
    let p = PathBuf::from(&path);
    Ok(p.canonicalize()
        .unwrap_or(p)
        .to_string_lossy()
        .to_string())
}
